### ISSUE TITLE
[Bug]: Server crash when reaching maximum file watchers

### ISSUE THREAD

**Issue by MikeGold** [advplyr/audiobookshelf#3258]
### What happened?

I'm using audiobookshelf for way over a year now. I never had issues with it. Recently I've added a lot of books and yesterday I've noticed for the first time that the audiobookshelf is crashed. I can't reach the webpage. To fix it I had to manually restart the container (although it has option to autorestart). This morning it crashed again. I've checked the logs and it seems to be filled with errors like this:

[Watcher] Error: ENOSPC: System limit for number of file watchers reached, watch '/audiobooks/Terry Pratchett/Świat Dysku/Vol 26 Złodziej czasu/@eaDir/07 Złodziej czasu.mp3'

The errors were generating non stop. I have stopped watcher in the settings, restarted the container and it works so far without errors. But it's odd.

I'm using audiobookshelf in the docker on Synology NAS 920+

[2024-08-08.txt](https://github.com/user-attachments/files/16548470/2024-08-08.txt)

### What did you expect to happen?

I expected the page to load

### Steps to reproduce the issue

1. I did nothing unusual. It just run and when I checked it the following day page was not loading

### Audiobookshelf version

v2.12.1

### How are you running audiobookshelf?

Docker

### What OS is your Audiobookshelf server hosted from?

Other (list in "Additional Notes" box)

### If the issue is being seen in the UI, what browsers are you seeing the problem on?

Chrome

### Logs

```shell
Logs attached above
```

### Additional Notes

Synology NAS docker
Checked on multiple browsers both mobile and desktop, on multiple internet connections local/remote.

**Comment by Vito0912** [comment]
You've hit the maximum number of file watchers allowed. Fortunately, you can adjust this limit on most os's. 

See also: https://github.com/advplyr/audiobookshelf/issues/905

**Comment by MikeGold** [comment]
I've added the command on start of Synology, and still got the error. But when I run the task on it's own after it booted, I haven't received any errors (yet). So I guess it's ok. Still, is it ok that it crashes the whole Audiobookshelf when it runs out of watchers?

**Comment by nichwall** [comment]
Do you have anything in `crash-logs.txt`? ABS is probably crashing for a different reason

**Comment by MikeGold** [comment]
No such file in the logs. The log from yesterday is a bare minimum, not stating any crash or anything. The only errors I've noticed were those watchers.

**Comment by rpgdev** [comment]
I'm having the same issue.. I just started using abs a few weeks ago so I only have 21 books on it. If it's because of file watchers then it's quite a coincidence. My primary language is Spanish and this happened without me realizing it at first just finding the container crashed and unable to access it then "matching" authors with Spanish names.

**Comment by rpgdev** [comment]
This is my crash log:
```txt
{"timestamp":"2024-08-08 13:23:54.961","source":"Server.js:171","message":"[Server] Unhandled rejection: SequelizeDatabaseError: SQLITE_CORRUPT: database disk image is malformed, promise: Promise {\n  <rejected> Error\n      at Database.<anonymous> (/node_modules/sequelize/lib/dialects/sqlite/query.js:185:27)\n      at /node_modules/sequelize/lib/dialects/sqlite/query.js:183:50\n      at new Promise (<anonymous>)\n      at Query.run (/node_modules/sequelize/lib/dialects/sqlite/query.js:183:12)\n      at /node_modules/sequelize/lib/sequelize.js:315:28\n      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n      at async SQLiteQueryInterface.bulkUpdate (/node_modules/sequelize/lib/dialects/abstract/query-interface.js:366:12)\n      at async author.update (/node_modules/sequelize/lib/model.js:2010:28)\n      at async ApiRouter.match (/server/controllers/AuthorController.js:325:7) {\n    name: 'SequelizeDatabaseError',\n    parent: [Error: SQLITE_CORRUPT: database disk image is malformed] {\n      errno: 11,\n      code: 'SQLITE_CORRUPT',\n      sql: 'UPDATE `authors` SET `id`=$1,`name`=$2,`lastFirst`=$3,`asin`=$4,`description`=$5,`imagePath`=$6,`libraryId`=$7,`updatedAt`=$8 WHERE `id` = $9'\n    },\n    original: [Error: SQLITE_CORRUPT: database disk image is malformed] {\n      errno: 11,\n      code: 'SQLITE_CORRUPT',\n      sql: 'UPDATE `authors` SET `id`=$1,`name`=$2,`lastFirst`=$3,`asin`=$4,`description`=$5,`imagePath`=$6,`libraryId`=$7,`updatedAt`=$8 WHERE `id` = $9'\n    },\n    sql: 'UPDATE `authors` SET `id`=$1,`name`=$2,`lastFirst`=$3,`asin`=$4,`description`=$5,`imagePath`=$6,`libraryId`=$7,`updatedAt`=$8 WHERE `id` = $9',\n    parameters: {}\n  }\n}","levelName":"FATAL","level":5}
```

**Comment by Scarprovider** [comment]
Same here, running on a Synology docker for more than 6 months without issue. Today just crashed, no upgrade involve,  without any reason and showed Unhealthy status in Portainer. A simple restart and ABS is running great again.  

Error text:

[AudioBookShelf] [🔴 Down] Request failed with status code 524

**Comment by glacasa** [comment]
Same here !
I run audiobookshelf in Docker, in a VPS. No problem for months, and then 2 crashes in the last 2 days.
The app just stops responding, nothing loads until I get a timeout. No error in the logs. 

I'm also on version v2.12.1

**Comment by Vito0912** [comment]
> Same here ! I run audiobookshelf in Docker, in a VPS. No problem for months, and then 2 crashes in the last 2 days. The app just stops responding, nothing loads until I get a timeout. No error in the logs.
> 
> I'm also on version v2.12.1

Do your CPU spikes during the downtime?

**Comment by glacasa** [comment]
> Do your CPU spikes during the downtime?

I can't tell because I don't monitor the server cpu, sorry 😅

But after looking [this other issue](https://github.com/advplyr/audiobookshelf/issues/3251), it seems I have the same problem : some people noticed ABS crashed after 24 hours, and I can confirm the second crash happened 24 hours after the last restart

**Comment by MikeGold** [comment]
For me it was also 2 crashes in the last 48 hours and no issues whatsoever in the past. My CPU does not spike and I never have downtimes as the same NAS is running a quite massive Plex server and about 10 other docker containers. Everything is incredibly stable and never freezes. It's the only docker container that I ever seen crashing. Maybe I'm just lucky...

**Comment by nikolausmoll** [comment]
Same here... didn't change anything, had to restart ABS yesterday and after about 24h right now again.

**Comment by bloodscript** [comment]
I today had this issue the second time, approximately 24 Hours After Last occurance. I looked into the Logs and found nothing that made me think that I found the issue. Will provide the Logs tomorrow. Running in Container on Pi for 6 Montag without any issues. Watchtower may have upgraded the Image...

**Comment by rweatherly** [comment]
same issue here. after the lates update (skipped 2.12.0 went directly to 2.12.1. haven't added any books, and same 2 users using it daily and the container has crashed each day now since the update

**Comment by ext4xfs** [comment]
> > Same here ! I run audiobookshelf in Docker, in a VPS. No problem for months, and then 2 crashes in the last 2 days. The app just stops responding, nothing loads until I get a timeout. No error in the logs.
> > I'm also on version v2.12.1
> 
> Do your CPU spikes during the downtime?

Mine was at 99% of one cpu, which is how I found that it was not responding. It has crashed twice now.

**Comment by DarkVirtue** [comment]
I think I'm having this issue also since updating to 2.12.1. The container stops responding to web requests almost exactly 24 hours after it was started.

**Comment by rweatherly** [comment]
> I think I'm having this issue also since updating to 2.12.1. The container stops responding to web requests almost exactly 24 hours after it was started.

im seeing the same thing, it was 24 hours from the restart yesterday till when it crashed again today

**Comment by daneroo** [comment]
Similar issue, 2 restarts in 2 days. Never happened before 2.12.1.
- container is still running, but not responding to web requests
- restart fixed it.

(docker compose on ubuntu 22.04)

Would love to help diagnose this; Let us know how to help to resolve  this.

**Comment by lyguang90716** [comment]
I think I'm having this issue also since updating to 2.12.1

**Comment by nichwall** [comment]
Release 2.12.2 released earlier today reverted the new Memory store, so the server should not stop responding after 24 hours

**Comment by AemonCao** [comment]
After updating to 2.12.2 I can't start the container.

```log
text  error  warn  system  array  login  

    config: {
      transitional: [Object],
      adapter: [Function: httpAdapter],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      maxBodyLength: -1,
      env: [Object],
      validateStatus: [Function: validateStatus],
      headers: [Object],
      method: 'get',
      url: 'https://api.github.com/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
      data: undefined
    },
    request: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: true,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: true,
      maxRequestsOnConnectionReached: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      strictContentLength: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      _closed: true,
      socket: [TLSSocket],
      _header: 'GET /repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1 HTTP/1.1\r\n' +
        'Accept: application/vnd.github.v3+json\r\n' +
        'User-Agent: axios\r\n' +
        'Host: api.github.com\r\n' +
        'Connection: keep-alive\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: nop],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      joinDuplicateHeaders: undefined,
      path: '/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
      _ended: true,
      res: [IncomingMessage],
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'api.github.com',
      protocol: 'https:',
      _redirectable: [Writable],
      [Symbol(shapeMode)]: false,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype],
      [Symbol(errored)]: null,
      [Symbol(kHighWaterMark)]: 16384,
      [Symbol(kRejectNonStandardBodyWrites)]: false,
      [Symbol(kUniqueHeaders)]: null
    },
    response: {
      status: 403,
      statusText: 'rate limit exceeded',
      headers: [Object],
      config: [Object],
      request: [ClientRequest],
      data: [Object]
    }
  }
} (Server.js:170)
```

**Comment by Klutronic** [comment]
I am also unable to start the container after upgrading to 2.12.2

audiobookshelf-1  | Config /config /metadata
audiobookshelf-1  | [2024-08-09 05:09:22.151] INFO: === Starting Server ===
audiobookshelf-1  | [2024-08-09 05:09:22.164] INFO: [Server] Init v2.12.2
audiobookshelf-1  | [2024-08-09 05:09:22.165] INFO: [Server] Node.js Version: v20.16.0
audiobookshelf-1  | [2024-08-09 05:09:22.165] INFO: [Server] Platform: linux
audiobookshelf-1  | [2024-08-09 05:09:22.165] INFO: [Server] Arch: x64
audiobookshelf-1  | [2024-08-09 05:09:22.259] INFO: [BinaryManager] ffmpeg not found or version too old
audiobookshelf-1  | [2024-08-09 05:09:22.311] INFO: [BinaryManager] ffprobe not found or version too old
audiobookshelf-1  | [2024-08-09 05:09:22.311] INFO: [BinaryManager] unicode not found or version too old
audiobookshelf-1  | [2024-08-09 05:09:22.312] INFO: [BinaryManager] Installing binaries: ffmpeg, ffprobe, unicode
audiobookshelf-1  | [2024-08-09 05:09:27.379] ERROR: [GithubAssetDownloader] Error downloading or extracting files: getaddrinfo EAI_AGAIN api.github.com (BinaryManager.js:99)
audiobookshelf-1  | [2024-08-09 05:09:27.385] FATAL: [Server] Unhandled rejection: Error: getaddrinfo EAI_AGAIN api.github.com, promise: Promise {
audiobookshelf-1  |   <rejected> AxiosError: getaddrinfo EAI_AGAIN api.github.com
audiobookshelf-1  |       at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:120:26) {
audiobookshelf-1  |     hostname: 'api.github.com',
audiobookshelf-1  |     syscall: 'getaddrinfo',
audiobookshelf-1  |     code: 'EAI_AGAIN',
audiobookshelf-1  |     errno: -3001,
audiobookshelf-1  |     config: {
audiobookshelf-1  |       transitional: [Object],
audiobookshelf-1  |       adapter: [Function: httpAdapter],
audiobookshelf-1  |       transformRequest: [Array],
audiobookshelf-1  |       transformResponse: [Array],
audiobookshelf-1  |       timeout: 0,
audiobookshelf-1  |       xsrfCookieName: 'XSRF-TOKEN',
audiobookshelf-1  |       xsrfHeaderName: 'X-XSRF-TOKEN',
audiobookshelf-1  |       maxContentLength: -1,
audiobookshelf-1  |       maxBodyLength: -1,
audiobookshelf-1  |       env: [Object],
audiobookshelf-1  |       validateStatus: [Function: validateStatus],
audiobookshelf-1  |       headers: [Object],
audiobookshelf-1  |       method: 'get',
audiobookshelf-1  |       url: 'https://api.github.com/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
audiobookshelf-1  |       data: undefined
audiobookshelf-1  |     },
audiobookshelf-1  |     request: Writable {
audiobookshelf-1  |       _events: [Object],
audiobookshelf-1  |       _writableState: [WritableState],
audiobookshelf-1  |       _maxListeners: undefined,
audiobookshelf-1  |       _options: [Object],
audiobookshelf-1  |       _ended: true,
audiobookshelf-1  |       _ending: true,
audiobookshelf-1  |       _redirectCount: 0,
audiobookshelf-1  |       _redirects: [],
audiobookshelf-1  |       _requestBodyLength: 0,
audiobookshelf-1  |       _requestBodyBuffers: [],
audiobookshelf-1  |       _eventsCount: 3,
audiobookshelf-1  |       _onNativeResponse: [Function (anonymous)],
audiobookshelf-1  |       _currentRequest: [ClientRequest],
audiobookshelf-1  |       _currentUrl: 'https://api.github.com/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
audiobookshelf-1  |       [Symbol(shapeMode)]: true,
audiobookshelf-1  |       [Symbol(kCapture)]: false
audiobookshelf-1  |     }
audiobookshelf-1  |   }
audiobookshelf-1  | } (Server.js:170)
audiobookshelf-1 exited with code 0

**Comment by rpgdev** [comment]
Updated to 2.12.2 and I see no issues on the logs so far :crossed_fingers:

**Comment by nikolausmoll** [comment]
@rpgdev: When did you update to 2.12.2?

**Comment by rpgdev** [comment]
@nikolausmoll just now.

**Comment by nikolausmoll** [comment]
and why should there be any issues then?

**Comment by SyndicatedPillbug** [comment]
Glad this isn't just me. I keep getting a "Warning" on my Container Manager in Synology. The container hasn't fully crashed, but it's not responding. This has happened three times in the last four or so days. I have to restart the container. Then it works for a while, but I inevitably have to go back and restart it again.

**Comment by MikeGold** [comment]
So I've mentioned yesterday that I've turned off the watchers and my logs were absolutely error free since. Today, about 24 hours since the last restart, my audiobookshelf had a fullblown crash that forced the docker container to restart. Well, that's something different. No crash logs :D. 

Should I update to 2.12.2? Some said that it's not starting for them...

**Comment by AemonCao** [comment]
> After updating to 2.12.2 I can't start the container.更新到 2.12.2 后，我无法启动容器。
> 
> ```
> text  error  warn  system  array  login  
> 
>     config: {
>       transitional: [Object],
>       adapter: [Function: httpAdapter],
>       transformRequest: [Array],
>       transformResponse: [Array],
>       timeout: 0,
>       xsrfCookieName: 'XSRF-TOKEN',
>       xsrfHeaderName: 'X-XSRF-TOKEN',
>       maxContentLength: -1,
>       maxBodyLength: -1,
>       env: [Object],
>       validateStatus: [Function: validateStatus],
>       headers: [Object],
>       method: 'get',
>       url: 'https://api.github.com/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
>       data: undefined
>     },
>     request: ClientRequest {
>       _events: [Object: null prototype],
>       _eventsCount: 7,
>       _maxListeners: undefined,
>       outputData: [],
>       outputSize: 0,
>       writable: true,
>       destroyed: true,
>       _last: true,
>       chunkedEncoding: false,
>       shouldKeepAlive: true,
>       maxRequestsOnConnectionReached: false,
>       _defaultKeepAlive: true,
>       useChunkedEncodingByDefault: false,
>       sendDate: false,
>       _removedConnection: false,
>       _removedContLen: false,
>       _removedTE: false,
>       strictContentLength: false,
>       _contentLength: 0,
>       _hasBody: true,
>       _trailer: '',
>       finished: true,
>       _headerSent: true,
>       _closed: true,
>       socket: [TLSSocket],
>       _header: 'GET /repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1 HTTP/1.1\r\n' +
>         'Accept: application/vnd.github.v3+json\r\n' +
>         'User-Agent: axios\r\n' +
>         'Host: api.github.com\r\n' +
>         'Connection: keep-alive\r\n' +
>         '\r\n',
>       _keepAliveTimeout: 0,
>       _onPendingData: [Function: nop],
>       agent: [Agent],
>       socketPath: undefined,
>       method: 'GET',
>       maxHeaderSize: undefined,
>       insecureHTTPParser: undefined,
>       joinDuplicateHeaders: undefined,
>       path: '/repos/ffbinaries/ffbinaries-prebuilt/releases/tags/v5.1',
>       _ended: true,
>       res: [IncomingMessage],
>       aborted: false,
>       timeoutCb: null,
>       upgradeOrConnect: false,
>       parser: null,
>       maxHeadersCount: null,
>       reusedSocket: false,
>       host: 'api.github.com',
>       protocol: 'https:',
>       _redirectable: [Writable],
>       [Symbol(shapeMode)]: false,
>       [Symbol(kCapture)]: false,
>       [Symbol(kBytesWritten)]: 0,
>       [Symbol(kNeedDrain)]: false,
>       [Symbol(corked)]: 0,
>       [Symbol(kOutHeaders)]: [Object: null prototype],
>       [Symbol(errored)]: null,
>       [Symbol(kHighWaterMark)]: 16384,
>       [Symbol(kRejectNonStandardBodyWrites)]: false,
>       [Symbol(kUniqueHeaders)]: null
>     },
>     response: {
>       status: 403,
>       statusText: 'rate limit exceeded',
>       headers: [Object],
>       config: [Object],
>       request: [ClientRequest],
>       data: [Object]
>     }
>   }
> } (Server.js:170)
> ```

Sorry, my reply may be off topic.

After an update, the container suddenly became inaccessible (it could run normally for a while after restart), but it became inaccessible again after a while (usually the next day). So today I updated to the latest version 2.12.2. Then the container became inaccessible.

After that, I had to downgrade to 2.11.0, which is still normal. I don’t know if the original inaccessibility problem will occur again tomorrow.

**Comment by ext4xfs** [comment]
Working fine on 2.12.2 here, no crashes, no high cpu

**Comment by SyndicatedPillbug** [comment]
![image](https://github.com/user-attachments/assets/294d910f-fb1d-481b-a77c-ce84fad19fdd)

Container crashed today. Not a weir freeze with a Warning like I saw before, but an actual crash notification. It auto-restarted fine, but...

**Comment by nikolausmoll** [comment]
I guess no one here testes 2.12.2 long enough to say that it does not crash after a day... i can tell tomorrow morning, if it crashed.

**Comment by advplyr** [comment]
@SyndicatedPillbug that warning isn't related to any crash. Are you able to get the actual crash log?
It would be in `/metadata/logs/crash_logs.txt`

**Comment by SyndicatedPillbug** [comment]
I don't see any crash_logs file, sadly. :/

**Comment by advplyr** [comment]
This issue got derailed into completely separate issues. This is specific to the server crashing when the system cannot assign anymore file watchers.
I will leave this open because we should try to handle this more gracefully and disable the watcher.

Any issues unrelated to file watchers should go in another thread.

**Comment by rpgdev** [comment]
@advplyr roger! I'll unwatch this issue since my issue was different, it just happened to coincide and the title of the bug described my situation (or someone linked it in a different one that was related).

**Comment by MikeGold** [comment]
@advplyr It was proven here, that my crashes were not related to the watchers errors, but were some other issue that does not leave trace in logs. Multiple other people had the same issue as I did (with semi crash after 24 hours) and reported it here. If you are planning to limit this bug report to just watchers, can you please point all of us to some new bug post that we can follow for that 24 hours crash? I don't think that anyone care about watchers, but the daily crashes are rather major issue for all of us.

**Comment by MikeGold** [comment]
@nikolausmoll Regardless everything else, please let us know in here tomorrow if the server crashed for you again.

**Comment by nikolausmoll** [comment]
I will. Didn't crash so far (17h runtime)
![grafik](https://github.com/user-attachments/assets/a543d298-6849-4a0e-a809-dcb19c9f43b9)

**Comment by MikeGold** [comment]
I believe that most appropriate place for the issue we all experienced is this bug: #3251 
So it may be better to move there.

**Comment by nikolausmoll** [comment]
agreed

**Comment by advplyr** [comment]
Ah okay thanks. I see the watcher isn't causing a crash, just an error message like it should. 

The 24h server locking up issue was the MemoryStore fixed in v2.12.2 reported in #3251.
Then there was a server crash issue happening that was fixed in v2.12.3.

Thanks to everyone reporting the issues
